File: src-tauri/src/commands/shots.rs
use crate::db::{generate_id, DbState};
use crate::media;
use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, State};

#[derive(Debug, Serialize, Clone)]
pub struct ShotWithUrls {
    pub id: String,
    pub storyboard_id: String,
    pub order: i64,
    pub duration: i64,
    pub image_prompt: Option<String>,
    pub image_path: Option<String>,
    pub video_prompt: Option<String>,
    pub video_path: Option<String>,
    pub status: String,
    pub created_at: String,
    pub updated_at: String,
    pub image_url: Option<String>,
    pub video_url: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct ShotUpdates {
    pub duration: Option<i64>,
    pub image_prompt: Option<String>,
    pub video_prompt: Option<String>,
    pub status: Option<String>,
}

/// Build an absolute file path for a media file.
/// Cache-busting timestamps are added on the TypeScript side via convertFileSrc.
fn make_media_url(app: &AppHandle, relative_path: &str) -> String {
    let base = media::get_media_base_dir(app);
    base.join(relative_path).to_string_lossy().into_owned()
}

/// Query a single shot and return it with URLs resolved.
fn query_shot_with_urls(
    conn: &rusqlite::Connection,
    app: &AppHandle,
    shot_id: &str,
) -> Result<ShotWithUrls, String> {
    let mut stmt = conn
        .prepare(
            r#"SELECT id, storyboard_id, "order", duration, image_prompt, image_path,
                      video_prompt, video_path, status, created_at, updated_at
               FROM shots WHERE id = ?1"#,
        )
        .map_err(|e| e.to_string())?;

    let shot = stmt
        .query_row(params![shot_id], |row| {
            let image_path: Option<String> = row.get(5)?;
            let video_path: Option<String> = row.get(7)?;

            let image_url = image_path.as_ref().map(|p| make_media_url(app, p));
            let video_url = video_path.as_ref().map(|p| make_media_url(app, p));

            Ok(ShotWithUrls {
                id: row.get(0)?,
                storyboard_id: row.get(1)?,
                order: row.get(2)?,
                duration: row.get(3)?,
                image_prompt: row.get(4)?,
                image_path,
                video_prompt: row.get(6)?,
                video_path,
                status: row.get(8)?,
                created_at: row.get(9)?,
                updated_at: row.get(10)?,
                image_url,
                video_url,
            })
        })
        .map_err(|e| e.to_string())?;

    Ok(shot)
}

#[tauri::command]
pub fn get_shots(
    storyboard_id: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<Vec<ShotWithUrls>, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    let mut stmt = conn
        .prepare(
            r#"SELECT id, storyboard_id, "order", duration, image_prompt, image_path,
                      video_prompt, video_path, status, created_at, updated_at
               FROM shots WHERE storyboard_id = ?1 ORDER BY "order" ASC"#,
        )
        .map_err(|e| e.to_string())?;

    let shots = stmt
        .query_map(params![storyboard_id], |row| {
            let image_path: Option<String> = row.get(5)?;
            let video_path: Option<String> = row.get(7)?;

            let image_url = image_path.as_ref().map(|p| make_media_url(&app, p));
            let video_url = video_path.as_ref().map(|p| make_media_url(&app, p));

            Ok(ShotWithUrls {
                id: row.get(0)?,
                storyboard_id: row.get(1)?,
                order: row.get(2)?,
                duration: row.get(3)?,
                image_prompt: row.get(4)?,
                image_path,
                video_prompt: row.get(6)?,
                video_path,
                status: row.get(8)?,
                created_at: row.get(9)?,
                updated_at: row.get(10)?,
                image_url,
                video_url,
            })
        })
        .map_err(|e| e.to_string())?
        .filter_map(|r| r.ok())
        .collect();

    Ok(shots)
}

#[tauri::command]
pub fn create_shot(
    storyboard_id: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;
    let id = generate_id("shot");

    // Get next order
    let next_order: i64 = conn
        .query_row(
            r#"SELECT COALESCE(MAX("order"), 0) + 1 FROM shots WHERE storyboard_id = ?1"#,
            params![storyboard_id],
            |row| row.get(0),
        )
        .map_err(|e| e.to_string())?;

    conn.execute(
        r#"INSERT INTO shots (id, storyboard_id, "order") VALUES (?1, ?2, ?3)"#,
        params![id, storyboard_id, next_order],
    )
    .map_err(|e| e.to_string())?;

    query_shot_with_urls(&conn, &app, &id)
}

#[tauri::command]
pub fn update_shot(
    id: String,
    updates_json: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let updates: ShotUpdates =
        serde_json::from_str(&updates_json).map_err(|e| format!("Invalid updates JSON: {}", e))?;

    let conn = state.0.lock().map_err(|e| e.to_string())?;

    let mut set_clauses = Vec::new();
    let mut param_values: Vec<Box<dyn rusqlite::types::ToSql>> = Vec::new();

    if let Some(duration) = updates.duration {
        set_clauses.push("duration = ?".to_string());
        param_values.push(Box::new(duration));
    }
    if let Some(ref image_prompt) = updates.image_prompt {
        set_clauses.push("image_prompt = ?".to_string());
        param_values.push(Box::new(image_prompt.clone()));
    }
    if let Some(ref video_prompt) = updates.video_prompt {
        set_clauses.push("video_prompt = ?".to_string());
        param_values.push(Box::new(video_prompt.clone()));
    }
    if let Some(ref status) = updates.status {
        set_clauses.push("status = ?".to_string());
        param_values.push(Box::new(status.clone()));
    }

    if !set_clauses.is_empty() {
        set_clauses.push("updated_at = CURRENT_TIMESTAMP".to_string());
        param_values.push(Box::new(id.clone()));

        let sql = format!(
            "UPDATE shots SET {} WHERE id = ?",
            set_clauses.join(", ")
        );

        let params_refs: Vec<&dyn rusqlite::types::ToSql> =
            param_values.iter().map(|p| p.as_ref()).collect();

        conn.execute(&sql, params_refs.as_slice())
            .map_err(|e| e.to_string())?;
    }

    query_shot_with_urls(&conn, &app, &id)
}

#[tauri::command]
pub fn delete_shot(
    id: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<bool, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Get shot to find media paths
    let shot_data: Option<(Option<String>, Option<String>)> = conn
        .query_row(
            "SELECT image_path, video_path FROM shots WHERE id = ?1",
            params![id],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )
        .ok();

    if let Some((image_path, video_path)) = shot_data {
        if let Some(path) = image_path {
            media::delete_media(&app, &path);
        }
        if let Some(path) = video_path {
            media::delete_media(&app, &path);
        }
    }

    // Delete version images and mask images
    media::delete_version_images(&app, &id);
    media::delete_mask_images(&app, &id);

    // Delete from DB (cascades to image_versions and timeline_edits)
    let changes = conn
        .execute("DELETE FROM shots WHERE id = ?1", params![id])
        .map_err(|e| e.to_string())?;

    Ok(changes > 0)
}

#[tauri::command]
pub fn reorder_shots(
    storyboard_id: String,
    shot_ids: Vec<String>,
    state: State<'_, DbState>,
) -> Result<(), String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Use a savepoint for the transaction
    let tx = conn
        .unchecked_transaction()
        .map_err(|e| e.to_string())?;

    for (index, shot_id) in shot_ids.iter().enumerate() {
        tx.execute(
            r#"UPDATE shots SET "order" = ?1 WHERE id = ?2 AND storyboard_id = ?3"#,
            params![index as i64 + 1, shot_id, storyboard_id],
        )
        .map_err(|e| e.to_string())?;
    }

    tx.commit().map_err(|e| e.to_string())?;
    Ok(())
}

#[tauri::command]
pub fn save_shot_image(
    id: String,
    base64_data_url: String,
    prompt: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Get existing shot to delete old media
    let existing: Option<(Option<String>, Option<String>)> = conn
        .query_row(
            "SELECT image_path, video_path FROM shots WHERE id = ?1",
            params![id],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )
        .ok();

    if let Some((old_image, old_video)) = &existing {
        if let Some(path) = old_image {
            media::delete_media(&app, path);
        }
        if let Some(path) = old_video {
            media::delete_media(&app, path);
        }
    }

    // Save new image
    let image_path = media::save_image(&app, &id, &base64_data_url)?;

    // Update shot in DB: set image, reset video
    conn.execute(
        "UPDATE shots SET image_path = ?1, image_prompt = ?2, video_path = NULL, status = 'pending', updated_at = CURRENT_TIMESTAMP WHERE id = ?3",
        params![image_path, prompt, id],
    )
    .map_err(|e| e.to_string())?;

    query_shot_with_urls(&conn, &app, &id)
}

#[tauri::command]
pub fn save_shot_video(
    id: String,
    base64_data_url: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Delete old video if exists
    let old_video: Option<String> = conn
        .query_row(
            "SELECT video_path FROM shots WHERE id = ?1",
            params![id],
            |row| row.get(0),
        )
        .ok()
        .flatten();

    if let Some(path) = old_video {
        media::delete_media(&app, &path);
    }

    // Save new video
    let video_path = media::save_video(&app, &id, &base64_data_url)?;

    // Update shot in DB
    conn.execute(
        "UPDATE shots SET video_path = ?1, status = 'complete', updated_at = CURRENT_TIMESTAMP WHERE id = ?2",
        params![video_path, id],
    )
    .map_err(|e| e.to_string())?;

    query_shot_with_urls(&conn, &app, &id)
}

#[tauri::command]
pub fn get_shot_image_base64(
    shot_id: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<Option<String>, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    let image_path: Option<String> = conn
        .query_row(
            "SELECT image_path FROM shots WHERE id = ?1",
            params![shot_id],
            |row| row.get(0),
        )
        .ok()
        .flatten();

    match image_path {
        Some(path) => media::get_image_as_base64(&app, &path),
        None => Ok(None),
    }
}

#[tauri::command]
pub fn copy_image_from_shot(
    target_shot_id: String,
    source_shot_id: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Get source shot's image
    let source: Option<(Option<String>, Option<String>, i64)> = conn
        .query_row(
            r#"SELECT image_path, image_prompt, "order" FROM shots WHERE id = ?1"#,
            params![source_shot_id],
            |row| Ok((row.get(0)?, row.get(1)?, row.get(2)?)),
        )
        .ok();

    let (source_image_path, source_prompt, source_order) =
        source.ok_or("Source shot not found")?;

    let source_image_path =
        source_image_path.ok_or("Source shot has no image")?;

    // Read source image as base64
    let image_base64 = media::get_image_as_base64(&app, &source_image_path)?
        .ok_or("Failed to read source image")?;

    // Delete old media from target
    let target: Option<(Option<String>, Option<String>)> = conn
        .query_row(
            "SELECT image_path, video_path FROM shots WHERE id = ?1",
            params![target_shot_id],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )
        .ok();

    if let Some((old_image, old_video)) = &target {
        if let Some(path) = old_image {
            media::delete_media(&app, path);
        }
        if let Some(path) = old_video {
            media::delete_media(&app, path);
        }
    }

    // Save copied image
    let image_path = media::save_image(&app, &target_shot_id, &image_base64)?;

    // Build prompt text
    let prompt_text = source_prompt
        .map(|_| format!("Copied from Shot #{}", source_order))
        .or_else(|| Some(format!("Copied from Shot #{}", source_order)));

    // Update target shot
    conn.execute(
        "UPDATE shots SET image_path = ?1, image_prompt = ?2, video_path = NULL, status = 'pending', updated_at = CURRENT_TIMESTAMP WHERE id = ?3",
        params![image_path, prompt_text, target_shot_id],
    )
    .map_err(|e| e.to_string())?;

    query_shot_with_urls(&conn, &app, &target_shot_id)
}

#[tauri::command]
pub fn save_and_complete_video(
    shot_id: String,
    video_data: Vec<u8>,
    mime_type: String,
    app: AppHandle,
    state: State<'_, DbState>,
) -> Result<ShotWithUrls, String> {
    let conn = state.0.lock().map_err(|e| e.to_string())?;

    // Delete old video if exists
    let old_video: Option<String> = conn
        .query_row(
            "SELECT video_path FROM shots WHERE id = ?1",
            params![shot_id],
            |row| row.get(0),
        )
        .ok()
        .flatten();

    if let Some(path) = old_video {
        media::delete_media(&app, &path);
    }

    // Save raw video bytes
    let video_path = media::save_video_blob(&app, &shot_id, &video_data, &mime_type)?;

    // Update shot in DB
    conn.execute(
        "UPDATE shots SET video_path = ?1, status = 'complete', updated_at = CURRENT_TIMESTAMP WHERE id = ?2",
        params![video_path, shot_id],
    )
    .map_err(|e| e.to_string())?;

    query_shot_with_urls(&conn, &app, &shot_id)
}

File: electron/commands/shots.ts
import path from "node:path";
import type { DatabaseSync } from "node:sqlite";
import { generateId } from "../db";
import {
  saveImage,
  saveEndFrame,
  saveVideo,
  saveVideoBlob,
  getImageAsBase64,
  deleteMedia,
  deleteVersionImages,
  deleteMaskImages,
} from "../media-files";

export interface ShotWithUrls {
  id: string;
  storyboard_id: string;
  order: number;
  duration: number;
  image_prompt: string | null;
  image_path: string | null;
  end_frame_path: string | null;
  video_prompt: string | null;
  video_path: string | null;
  status: string;
  created_at: string;
  updated_at: string;
  image_url: string | null;
  end_frame_url: string | null;
  video_url: string | null;
}

/** Whitelisted, updatable fields, mirroring the Rust `ShotUpdates` struct. */
const UPDATABLE_FIELDS = ["duration", "image_prompt", "video_prompt", "status"] as const;

interface ShotRow {
  id: string;
  storyboard_id: string;
  order: number;
  duration: number;
  image_prompt: string | null;
  image_path: string | null;
  end_frame_path: string | null;
  video_prompt: string | null;
  video_path: string | null;
  status: string;
  created_at: string;
  updated_at: string;
}

const SHOT_COLUMNS = `id, storyboard_id, "order", duration, image_prompt, image_path,
                      end_frame_path, video_prompt, video_path, status, created_at, updated_at`;

/** Ported shot commands; names and JSON shapes match the retired Rust backend's commands/shots.rs. */
export function createShotCommands(db: DatabaseSync, mediaDir: string) {
  /**
   * Build an absolute file path for a media file, mirroring Rust make_media_url
   * (media base dir joined onto the relative DB path). Cache-busting timestamps
   * are added on the TypeScript side via convertFileSrc.
   */
  function makeMediaUrl(relativePath: string | null): string | null {
    return relativePath === null ? null : path.join(mediaDir, relativePath);
  }

  function rowToShot(row: ShotRow): ShotWithUrls {
    return {
      id: row.id,
      storyboard_id: row.storyboard_id,
      order: row.order,
      duration: row.duration,
      image_prompt: row.image_prompt,
      image_path: row.image_path,
      end_frame_path: row.end_frame_path,
      video_prompt: row.video_prompt,
      video_path: row.video_path,
      status: row.status,
      created_at: row.created_at,
      updated_at: row.updated_at,
      image_url: makeMediaUrl(row.image_path),
      end_frame_url: makeMediaUrl(row.end_frame_path),
      video_url: makeMediaUrl(row.video_path),
    };
  }

  /** Query a single shot and return it with URLs resolved. */
  function queryShotWithUrls(shotId: string): ShotWithUrls {
    const row = db
      .prepare(`SELECT ${SHOT_COLUMNS} FROM shots WHERE id = ?`)
      .get(shotId) as ShotRow | undefined;
    if (!row) {
      // Deliberate deviation: Rust's query_row surfaces rusqlite's incidental
      // "Query returned no rows" here; no caller matches on the text.
      throw new Error("Shot not found");
    }
    return rowToShot(row);
  }

  function getShots(storyboardId: string): ShotWithUrls[] {
    const rows = db
      .prepare(`SELECT ${SHOT_COLUMNS} FROM shots WHERE storyboard_id = ? ORDER BY "order" ASC`)
      .all(storyboardId) as unknown as ShotRow[];
    return rows.map(rowToShot);
  }

  function createShot(storyboardId: string): ShotWithUrls {
    const id = generateId("shot");

    const { next_order: nextOrder } = db
      .prepare(
        `SELECT COALESCE(MAX("order"), 0) + 1 AS next_order FROM shots WHERE storyboard_id = ?`
      )
      .get(storyboardId) as { next_order: number };

    db.prepare(`INSERT INTO shots (id, storyboard_id, "order") VALUES (?, ?, ?)`).run(
      id,
      storyboardId,
      nextOrder
    );

    return queryShotWithUrls(id);
  }

  function updateShot(id: string, updatesJson: string): ShotWithUrls {
    let updates: Record<string, unknown>;
    try {
      const parsed = JSON.parse(updatesJson);
      if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
        throw new Error("expected a JSON object");
      }
      updates = parsed as Record<string, unknown>;
    } catch (e) {
      throw new Error(`Invalid updates JSON: ${e instanceof Error ? e.message : String(e)}`);
    }

    const setClauses: string[] = [];
    const paramValues: unknown[] = [];

    for (const field of UPDATABLE_FIELDS) {
      // Mirror serde's Option semantics: an absent or null field is skipped.
      const value = updates[field];
      if (value !== undefined && value !== null) {
        setClauses.push(`${field} = ?`);
        paramValues.push(value);
      }
    }

    if (setClauses.length > 0) {
      setClauses.push("updated_at = CURRENT_TIMESTAMP");
      paramValues.push(id);
      const sql = `UPDATE shots SET ${setClauses.join(", ")} WHERE id = ?`;
      db.prepare(sql).run(...(paramValues as never[]));
    }

    return queryShotWithUrls(id);
  }

  function deleteShot(id: string): boolean {
    const shot = db
      .prepare("SELECT image_path, end_frame_path, video_path FROM shots WHERE id = ?")
      .get(id) as
      | { image_path: string | null; end_frame_path: string | null; video_path: string | null }
      | undefined;

    if (shot) {
      if (shot.image_path) deleteMedia(mediaDir, shot.image_path);
      if (shot.end_frame_path) deleteMedia(mediaDir, shot.end_frame_path);
      if (shot.video_path) deleteMedia(mediaDir, shot.video_path);
    }

    deleteVersionImages(mediaDir, id);
    deleteMaskImages(mediaDir, id);

    const { changes } = db.prepare("DELETE FROM shots WHERE id = ?").run(id);
    return Number(changes) > 0;
  }

  function reorderShots(storyboardId: string, shotIds: string[]): void {
    db.exec("BEGIN");
    try {
      const stmt = db.prepare(
        `UPDATE shots SET "order" = ? WHERE id = ? AND storyboard_id = ?`
      );
      shotIds.forEach((shotId, index) => {
        stmt.run(index + 1, shotId, storyboardId);
      });
      db.exec("COMMIT");
    } catch (e) {
      db.exec("ROLLBACK");
      throw e;
    }
  }

  function saveShotImage(id: string, base64DataUrl: string, prompt: string): ShotWithUrls {
    const existing = db
      .prepare("SELECT image_path, video_path FROM shots WHERE id = ?")
      .get(id) as { image_path: string | null; video_path: string | null } | undefined;

    if (existing) {
      if (existing.image_path) deleteMedia(mediaDir, existing.image_path);
      if (existing.video_path) deleteMedia(mediaDir, existing.video_path);
    }

    const imagePath = saveImage(mediaDir, id, base64DataUrl);

    db.prepare(
      "UPDATE shots SET image_path = ?, image_prompt = ?, video_path = NULL, status = 'pending', updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(imagePath, prompt, id);

    return queryShotWithUrls(id);
  }

  function saveShotEndFrame(id: string, base64DataUrl: string): ShotWithUrls {
    const oldEndFrame = (
      db.prepare("SELECT end_frame_path FROM shots WHERE id = ?").get(id) as
        | { end_frame_path: string | null }
        | undefined
    )?.end_frame_path;

    if (oldEndFrame) deleteMedia(mediaDir, oldEndFrame);

    const endFramePath = saveEndFrame(mediaDir, id, base64DataUrl);

    db.prepare(
      "UPDATE shots SET end_frame_path = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(endFramePath, id);

    return queryShotWithUrls(id);
  }

  function clearShotEndFrame(id: string): ShotWithUrls {
    const oldEndFrame = (
      db.prepare("SELECT end_frame_path FROM shots WHERE id = ?").get(id) as
        | { end_frame_path: string | null }
        | undefined
    )?.end_frame_path;

    if (oldEndFrame) deleteMedia(mediaDir, oldEndFrame);

    db.prepare(
      "UPDATE shots SET end_frame_path = NULL, updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(id);

    return queryShotWithUrls(id);
  }

  function saveShotVideo(id: string, base64DataUrl: string): ShotWithUrls {
    const oldVideo = (
      db.prepare("SELECT video_path FROM shots WHERE id = ?").get(id) as
        | { video_path: string | null }
        | undefined
    )?.video_path;

    if (oldVideo) deleteMedia(mediaDir, oldVideo);

    const videoPath = saveVideo(mediaDir, id, base64DataUrl);

    db.prepare(
      "UPDATE shots SET video_path = ?, status = 'complete', updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(videoPath, id);

    return queryShotWithUrls(id);
  }

  function getShotImageBase64(shotId: string): string | null {
    const imagePath = (
      db.prepare("SELECT image_path FROM shots WHERE id = ?").get(shotId) as
        | { image_path: string | null }
        | undefined
    )?.image_path;

    return imagePath ? getImageAsBase64(mediaDir, imagePath) : null;
  }

  function getShotEndFrameBase64(shotId: string): string | null {
    const endFramePath = (
      db.prepare("SELECT end_frame_path FROM shots WHERE id = ?").get(shotId) as
        | { end_frame_path: string | null }
        | undefined
    )?.end_frame_path;

    return endFramePath ? getImageAsBase64(mediaDir, endFramePath) : null;
  }

  function copyImageFromShot(targetShotId: string, sourceShotId: string): ShotWithUrls {
    const source = db
      .prepare(`SELECT image_path, image_prompt, "order" AS source_order FROM shots WHERE id = ?`)
      .get(sourceShotId) as
      | { image_path: string | null; image_prompt: string | null; source_order: number }
      | undefined;

    if (!source) {
      throw new Error("Source shot not found");
    }

    const sourceImagePath = source.image_path;
    if (!sourceImagePath) {
      throw new Error("Source shot has no image");
    }

    const imageBase64 = getImageAsBase64(mediaDir, sourceImagePath);
    if (imageBase64 === null) {
      throw new Error("Failed to read source image");
    }

    const target = db
      .prepare("SELECT image_path, video_path FROM shots WHERE id = ?")
      .get(targetShotId) as
      | { image_path: string | null; video_path: string | null }
      | undefined;

    if (target) {
      if (target.image_path) deleteMedia(mediaDir, target.image_path);
      if (target.video_path) deleteMedia(mediaDir, target.video_path);
    }

    const imagePath = saveImage(mediaDir, targetShotId, imageBase64);

    // Rust builds the same string on both map branches; the prompt is always set.
    const promptText = `Copied from Shot #${source.source_order}`;

    db.prepare(
      "UPDATE shots SET image_path = ?, image_prompt = ?, video_path = NULL, status = 'pending', updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(imagePath, promptText, targetShotId);

    return queryShotWithUrls(targetShotId);
  }

  function saveAndCompleteVideo(
    shotId: string,
    videoData: number[],
    mimeType: string
  ): ShotWithUrls {
    const oldVideo = (
      db.prepare("SELECT video_path FROM shots WHERE id = ?").get(shotId) as
        | { video_path: string | null }
        | undefined
    )?.video_path;

    if (oldVideo) deleteMedia(mediaDir, oldVideo);

    const videoPath = saveVideoBlob(mediaDir, shotId, Uint8Array.from(videoData), mimeType);

    db.prepare(
      "UPDATE shots SET video_path = ?, status = 'complete', updated_at = CURRENT_TIMESTAMP WHERE id = ?"
    ).run(videoPath, shotId);

    return queryShotWithUrls(shotId);
  }

  return {
    get_shots(args?: Record<string, unknown>): ShotWithUrls[] {
      return getShots(args!.storyboardId as string);
    },
    create_shot(args?: Record<string, unknown>): ShotWithUrls {
      return createShot(args!.storyboardId as string);
    },
    update_shot(args?: Record<string, unknown>): ShotWithUrls {
      return updateShot(args!.id as string, args!.updatesJson as string);
    },
    delete_shot(args?: Record<string, unknown>): boolean {
      return deleteShot(args!.id as string);
    },
    reorder_shots(args?: Record<string, unknown>): void {
      return reorderShots(args!.storyboardId as string, args!.shotIds as string[]);
    },
    save_shot_image(args?: Record<string, unknown>): ShotWithUrls {
      return saveShotImage(
        args!.id as string,
        args!.base64DataUrl as string,
        args!.prompt as string
      );
    },
    save_shot_end_frame(args?: Record<string, unknown>): ShotWithUrls {
      return saveShotEndFrame(args!.id as string, args!.base64DataUrl as string);
    },
    clear_shot_end_frame(args?: Record<string, unknown>): ShotWithUrls {
      return clearShotEndFrame(args!.id as string);
    },
    save_shot_video(args?: Record<string, unknown>): ShotWithUrls {
      return saveShotVideo(args!.id as string, args!.base64DataUrl as string);
    },
    get_shot_image_base64(args?: Record<string, unknown>): string | null {
      return getShotImageBase64(args!.shotId as string);
    },
    get_shot_end_frame_base64(args?: Record<string, unknown>): string | null {
      return getShotEndFrameBase64(args!.shotId as string);
    },
    copy_image_from_shot(args?: Record<string, unknown>): ShotWithUrls {
      return copyImageFromShot(
        args!.targetShotId as string,
        args!.sourceShotId as string
      );
    },
    save_and_complete_video(args?: Record<string, unknown>): ShotWithUrls {
      return saveAndCompleteVideo(
        args!.shotId as string,
        args!.videoData as number[],
        args!.mimeType as string
      );
    },
  };
}
